src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/ProjectConfigurationStateSynchronizer.cs

```diff
@@ -21,23 +21,54 @@
 
 namespace Microsoft.AspNetCore.Razor.LanguageServer;
 
+/// <summary>
```

Comment:
Love this detailed comment! Is the issue before that async batching could cause problems when doing a lookup for projectkey to file?

Reply:
The problem was that the project would get added early enough. Essentially, the async batching queue leveled the playing field so that everything happens in the same queue. However, the project add needs to happen before enqueueing an update. Otherwise, we get into bad behavior where a remove followed by an add (which happens when we delete and write a new file) causes files to be migrated to and from the misc project needlessly. @davidwengier wrote up a good explanation in https://github.com/dotnet/razor/issues/10306.

Reply:
For the record, I had a branch that kept the async batching queue, and just collapsed a Reset-then-Add to an Update, and it fixed the issue, so we could revisit this in future. At least we better understand the traps to look out for.

Also worth noting that even with the async batching queue, the integration tests pass. I noticed the issue due to some debug asserts in the `MoveDocument` method on `RazorProjectService`, and scouring the logs, but that was all caused by things happening between the reset and the add, and after the add the final state of things was all correct.

Reply:
Agreed. I actually think things worked with the async batching queue. It just did a bunch of unnecessary work that I fully agree that we don't want to do. I'd prefer deleting this type altogether rather than spending too much more time on it.

Reply:
I put up https://github.com/dotnet/razor/pull/10330 as an alternative to this. Curious for peoples thoughts.